### ISSUE TITLE
`FinallyGuard` included by multiple libraries; Leads to linking errors

### ISSUE THREAD

**Issue by jbfitb** [alexa/avs-device-sdk#1517]
## Briefly summarize your issue:
 - Compiling for iOS
 - Trying to link static libraries into a single executable
 - Getting linker errors:

```
duplicate symbol 'alexaClientSDK::avsCommon::utils::error::FinallyGuard::~FinallyGuard()' in:
    <redacted_path>(AlertScheduler.cpp.o)
    <redacted_path>(LocaleWakeWordsSetting.cpp.o)
duplicate symbol 'alexaClientSDK::avsCommon::utils::error::FinallyGuard::~FinallyGuard()' in:
    <redacted_path>(AlertScheduler.cpp.o)
    <redacted_path>(LocaleWakeWordsSetting.cpp.o)
duplicate symbol 'alexaClientSDK::avsCommon::utils::error::FinallyGuard::FinallyGuard(std::__1::function<void ()> const&)' in:
    <redacted_path>(AlertScheduler.cpp.o)
    <redacted_path>(LocaleWakeWordsSetting.cpp.o)
duplicate symbol 'alexaClientSDK::avsCommon::utils::error::FinallyGuard::FinallyGuard(std::__1::function<void ()> const&)' in:
    <redacted_path>(AlertScheduler.cpp.o)
    <redacted_path>(LocaleWakeWordsSetting.cpp.o)
duplicate symbol 'alexaClientSDK::avsCommon::utils::error::FinallyGuard::~FinallyGuard()' in:
    <redacted_path>(AlertScheduler.cpp.o)
    <redacted_path>(ReportStateHandler.cpp.o)
duplicate symbol 'alexaClientSDK::avsCommon::utils::error::FinallyGuard::~FinallyGuard()' in:
    <redacted_path>(AlertScheduler.cpp.o)
    <redacted_path>(ReportStateHandler.cpp.o)
duplicate symbol 'alexaClientSDK::avsCommon::utils::error::FinallyGuard::FinallyGuard(std::__1::function<void ()> const&)' in:
    <redacted_path>(AlertScheduler.cpp.o)
    <redacted_path>(ReportStateHandler.cpp.o)
duplicate symbol 'alexaClientSDK::avsCommon::utils::error::FinallyGuard::FinallyGuard(std::__1::function<void ()> const&)' in:
    <redacted_path>(AlertScheduler.cpp.o)
    <redacted_path>(ReportStateHandler.cpp.o)
ld: 8 duplicate symbols for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

## Research into the Issue
 - It looks like [`FinallyGuard` has its implementation in its header file](https://github.com/alexa/avs-device-sdk/blob/master/AVSCommon/Utils/include/AVSCommon/Utils/Error/FinallyGuard.h#L64)
 - Before 1.15, FinallyGuard was only used in [`AlertScheduler.cpp`](https://github.com/alexa/avs-device-sdk/blob/master/CapabilityAgents/Alerts/src/AlertScheduler.cpp) so this wasn't an issue.
- As of 1.15, `FinallyGuard` is used by three objects: `AlertScheduler.cpp.o`, `LocaleWakeWordsSetting.cpp.o`, and `ReportStateHandler.cpp.o`
 - The `FinallyGuard` symbols are now referenced in `libAVSSystem`, `libAlerts`, and `libDeviceSettings`
 - When linking two or more of these three libraries, there are errors due to duplicate symbols.

## Possible Solution

It seems that creating a `FinallyGuard.cpp` and removing the implementation from the header file fixes this issue:

`FinallyGuard.h`
```
#ifndef ALEXA_CLIENT_SDK_AVSCOMMON_UTILS_INCLUDE_AVSCOMMON_UTILS_ERROR_FINALLYGUARD_H_
#define ALEXA_CLIENT_SDK_AVSCOMMON_UTILS_INCLUDE_AVSCOMMON_UTILS_ERROR_FINALLYGUARD_H_

#include <functional>

namespace alexaClientSDK {
namespace avsCommon {
namespace utils {
namespace error {

/**
 * Define a class that can be used to run a function when the object goes out of scope.
 *
 * This simulates try-finally statements. The following structure:
 *
 * @code
 * try {
 *   <try_block>
 * } finally {
 *   <finally_block>
 * }
 * @endcode
 *
 * can be replaced by:
 *
 * @code
 *   FinallyGuard guard { []{ <finally_block> }};
 *   <try_block>
 * @endcode
 */
class FinallyGuard {
public:
    /**
     * Constructor.
     *
     * @param finallyFunction The function to be executed when the object goes out of scope.
     */
    FinallyGuard(const std::function<void()>& finallyFunction);

    /**
     * Destructor. Runs @c m_function during destruction.
     */
    ~FinallyGuard();

private:
    /// The function to be run when this object goes out of scope.
    std::function<void()> m_function;
};

}  // namespace error
}  // namespace utils
}  // namespace avsCommon
}  // namespace alexaClientSDK
#endif  // ALEXA_CLIENT_SDK_AVSCOMMON_UTILS_INCLUDE_AVSCOMMON_UTILS_ERROR_FINALLYGUARD_H_
```

`FinallyGuard.cpp`
```
#include "AVSCommon/Utils/Error/FinallyGuard.h"

namespace alexaClientSDK {
namespace avsCommon {
namespace utils {
namespace error {

FinallyGuard::FinallyGuard(const std::function<void()>& finallyFunction) : m_function{finallyFunction} {
}

FinallyGuard::~FinallyGuard() {
    if (m_function) {
        m_function();
    }
}

}  // namespace error
}  // namespace utils
}  // namespace avsCommon
}  // namespace alexaClientSDK
```

`AVSCommon/CMakeLists.txt`
```
+     Utils/src/FinallyGuard.cpp
```

**Comment by scotthea-amazon** [comment]
Hello @jbfitb,

This is definitely a bug.  Thank you for reporting this issue!

Best regards,
-SWH

**Comment by caleighatamazon** [comment]
Closing this issue for now. We're tracking the bug internally. Please feel free to re-open the issue if you have more questions or concerns.

**Comment by scotthea-amazon** [comment]
Hello @jbfitb,

This issue has been fixed in the 1.17 release.  Thank you again (and again) for your help!

Best regards,
-SWH
